### PR TITLE
Document which conversions happen in isSubtype

### PR DESCRIPTION
**Pull request by smillst**
String conversions should be handled via a method in AnnotatedTypeFactory. (See issue #715)

### CODE REVIEW

File: framework/src/org/checkerframework/framework/type/TypeHierarchy.java [hunk 1]
```diff
@@ -1,5 +1,7 @@
 package org.checkerframework.framework.type;
 
+import org.checkerframework.framework.util.AnnotatedTypes;
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/TypeHierarchy.java]: Probably no longer needed.

File: framework/src/org/checkerframework/framework/type/TypeHierarchy.java [hunk 2]
```diff
@@ -8,13 +10,58 @@
  */
 public interface TypeHierarchy {
 
+
     /**
-     * @return true if subtype {@literal <:} supertype for all qualifier hierarchies in the type system
+     * Returns true if {@code subtype} is a subtype of or convertible to {@code supertype}
+     * for all hierarchies present.  If the underlying Java type of {@code subtype} must be a subtype of
+     * or convertible to the underlying Java type of {@code supertype}, then the behavior of this method is undefined.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/TypeHierarchy.java]: I don't understand when the behavior is undefined. Usually the underlying type of `subtype` will be a subtype of the underlying type of `supertype`. Why would this be undefined?

File: framework/src/org/checkerframework/framework/type/TypeHierarchy.java [hunk 3]
```diff
@@ -8,13 +10,58 @@
...
+     *
+     * Ideally, types that require conversions would be converted before isSubtype is called, but instead, isSubtype
+     * performs some of these conversions.
+     *
+     * JLS 5.1 specifies 13 categories of conversions.  4 categories are converted in isSubtype:
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/TypeHierarchy.java]: I wouldn't start a sentence with the literal 4. Same with `1` below, where also the indentation is off.

File: framework/src/org/checkerframework/framework/type/TypeHierarchy.java [hunk 4]
```diff
@@ -8,13 +10,58 @@
...
+     * <ul>
+     * <li> Boxing conversions: isSubtype calls {@link AnnotatedTypes#asSuper}
+     *      which calls {@link AnnotatedTypeFactory#getBoxedType}
+     * <li> Unboxing conversions: isSubtype calls {@link AnnotatedTypes#asSuper}
+     *      which calls {@link AnnotatedTypeFactory#getUnboxedType}
+     * <li> Capture conversions:  Wildcards are treated as though they were converted to type variables
+     * <li> String conversions: Any type to String. Special case in {@link DefaultTypeHierarchy#visitDeclared_Declared}.
+     * </ul>
+     *
+     *  1 happens else where:
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/TypeHierarchy.java]: -> elsewhere

File: framework/src/org/checkerframework/framework/type/TypeHierarchy.java [hunk 5]
```diff
@@ -8,13 +10,58 @@
...
+     * <ul>
+     * <li> Unchecked conversions: Generic type to raw type.  Raw types are instantiated with bounds in
+     *      {@link AnnotatedTypeFactory#fromTypeTree} before is subtype is called
+     * </ul>
+     *
+     * 7 are not explicitly converted and are treated as though the types are actually subtypes.
+     * <ul>
+     * <li> Identity conversions: type to same type
+     * <li> Widening primitive conversions: primitive to primitive (no lose of information, byte to short for example)
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/TypeHierarchy.java]: lose -> loss

File: framework/src/org/checkerframework/framework/type/TypeHierarchy.java [hunk 6]
```diff
@@ -8,13 +10,58 @@
...
+     * <li> Narrowing reference conversions: Downcast
+     * <li> Value set conversions: floating-point value from one value set to another without changing its type.
+     * </ul>
+     *
+     * @param subtype   possible subtype
+     * @param supertype possible supertype
+     * @return true if {@code subtype} is a subtype of {@code supertype} for all hierarchies present.
      */
-    public boolean isSubtype(AnnotatedTypeMirror subtype, AnnotatedTypeMirror supertype);
+     boolean isSubtype(AnnotatedTypeMirror subtype, AnnotatedTypeMirror supertype);
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/type/TypeHierarchy.java]: Indentation was increased, also below.

**Comment by wmdietl** [comment]
This is a great improvement! Thanks for tracking down where the different conversions happen.
I only have a few small editorial comments. Can be merged afterwards.

File: framework/src/org/checkerframework/framework/type/TypeHierarchy.java (hunk 1 above)
Comment (smillst) [inline framework/src/org/checkerframework/framework/type/TypeHierarchy.java]: It's used in the Javadoc

File: framework/src/org/checkerframework/framework/type/TypeHierarchy.java (hunk 2 above)
Comment (smillst) [inline framework/src/org/checkerframework/framework/type/TypeHierarchy.java]: That should have read "If the underlying Java type is not a subtype..."

**Comment by smillst** [comment]
Rebased version commit in 92513e7.

**Pull request closed without merging**